javascript/react-js/react-testing-part-two.md

```diff
@@ -0,0 +1,155 @@
+
+#### submissions-list.test.jsx
+
+We notice there are two child components of `SubmissionsList`. One of them is from a package called `react-flip-move`. External Code. We'll mock it. 
+
+The other one is `Submission`. On a glance, there might not seem a reason to mock `Submission`, but if we dig deeper `Submission` has a lot going on. It has many child components of its own. And they all of them must have been tested beforehand. It's not ideal to render them all if just want to test `SubmissionsList`. Notice how we mock it:
+
+~~~javascript 
+jest.mock('../submission', () => ({ submission, isDashboardView }) => (
+  <>
+    <div data-test-id="submission">{submission.id}</div>
+    <div data-test-id="dashboard">{isDashboardView.toString()}</div>
+  </>
+));
+~~~
+
+We only render the bare minimum to realize the validity of the component we're testing. Next, we set up our props with fake data and mocked functions.
+
+Let's move towards our first assertion. Don't worry too much about the `ProjectSubmissionContext.Provider`. In the context of this test, it's purpose is to act as a route to pass in the `allSubmissionsPath` prop. We've already identified the three points of interest that we want to test. We divide them into three test suites for readability purposes using `describe`.
+
+In the first suite, we make some assertions if the user has a submssion and then some assertions if the user does not. The other suites follow a similar pattern.
+
+#### Exploring Further
+
+Feel free to flick through the other components and its tests. You'll see mocked functions. You'll see some unseen functions like `act`. You'll see custom render functions. It's fine if you don't understand it all, the goal is to gain familiarity.
+
+The other important thing to note is almost all the tests follow a certain pattern in terms of the way they're written. They follow the [Arrange-Act-Assert](http://wiki.c2.com/?ArrangeActAssert) pattern. It's a good idea to adopt the pattern sooner or later to make your tests more readable and ultimately better.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Read about the [secrets of act(...) api](https://github.com/mrdulin/react-act-examples/blob/master/sync.md). Especially pay attention to the last bit about testing asynchronous code. You won't be using it everyday, but it's nice to be familiar.
+
+2. We'll practice what we've learned with a simple exercise; a comment liker application.  
+    1. It's task is simple. Fetch dummy comments from [jsonplaceholder](https://jsonplaceholder.typicode.com/). Add a like and a dislike button to each comment. Display a counter for likes/dislikes. Once a button is clicked, increment the desired count and delete the comment (optional).
```

Comment:
```suggestion
    1. Its task is simple: fetch dummy comments from [jsonplaceholder](https://jsonplaceholder.typicode.com/). Add a like and a dislike button to each comment. Display a counter for likes/dislikes. Once a button is clicked, increment the desired count and delete the comment (optional).
```

Reply:
I think that this exercise would be more efficient if it was just about testing a component, rather than building a small project from scratch. Doing so feels out of scope to me considering the context of what the lesson is about.

Reply:
https://github.com/01zulfi/comment-liker-app

What do you think about this? It's pretty bare bones. Instead of the writing theproject from scratch, we provide them with all the code sans the tests...

Reply:
I'll mess around with the application a bit while I'm working through the onChange testing question and might have some additional feedback after that, but from a quick look at least I think that would probably be perfect. I definitely prefer this direction over having students build the project themselves.

One note: before we deploy this lesson we would probably want to add this application as something hosted directly under the TOP repo somehwere, I'll do some thinking on that side of things as well and let you know.